doc/developer/design/20231211_datadog_sink.md

```diff
@@ -0,0 +1,162 @@
+
+```sql
+CREATE TABLE metrics (
+    -- Required:
+    metric TEXT,
+    resources JSONB,
+    value FLOAT,
+    -- Optional:
+    type INT,
+    interval BIGINT,
+    tags TEXT[],
+    unit TEXT,
+    metadata: JSONB,
+);
+```
+
+And a view needs the same structure:
+```sql
+CREATE MATERIALIZED VIEW metrics AS
+    SELECT
+    -- Required:
+    DISTINCT(
+        jsonb_build_object(
+            'cluster_id', R.cluster_id,
+            'cluster_name', C.name,
+        )
+    ) as resources,
+    'cluster_replica_cpu_usage' as metric,
+    cpu_percent::float as value,
+    -- Optional:
+    0 as type,
+        120 as interval,
+    ARRAY['production', 'Materialize'] as tags,
+    'percent' as unit
+    FROM mz_internal.mz_cluster_replica_utilization U
+    JOIN mz_catalog.mz_cluster_replicas R ON (U.replica_id = R.id)
+    JOIN mz_catalog.mz_clusters C ON (R.cluster_id = C.id);
+```
+
+**NOTE**: Metrics and their values do not need to be unique for each timestamp, but the latest value will suppress the other in Datadog. It would be great if the sink could detect this during creation/execution and emit a warning if there are different values at the same point in time for a set of metrics and resources.
+
+Another distinction is that *resources* equal *labels* in the OpenMetrics definition. I kept *resources* to match Datadog’s API, but I like *labels* as a possible replacement.
+
+### OAuth Process
+
+Datadog recommends an OAuth process for their integrations and creating an API key on behalf of the user if the integration needs to write data into Datadog. The token does not have the scope to write data into Datadog, so creating the API key is necessary.
+
+The OAuth process requires additional logic in the console. The whole process is defined in the following sequence diagram:
+
+<details open>
+<summary>Sequence diagram</summary>
+<img width="2129" alt="Sequence Diagram" src="https://github.com/joacoc/materialize/assets/11491779/28dc14cb-a488-4e8b-8c89-ab69accfd858">
+</details>
+
+At the end of the flow, the console creates a secret containing the Datadog API key, connection, and sink. Although we lack an integrations page, a screen similar to what the mz displays at the end of the flow is more than enough for an MVP.
+
+### Datadog Sink Code
+
+The Datadog sink will be added as a new module at `storage/src/sink` and have its own implementation of `SinkRender`. The Datadog sink doesn't need to send retractions or manage a complex state. Shared code paths between the Kafka sink and the Datadog sink will surge and should be handled with consideration. E.g. adding a new sink could introduce changes in places like (`alter_compatible`)[https://github.com/MaterializeInc/materialize/blob/cc947745215d96e4d86ff507b6ac92311690d7d3/src/storage-types/src/sinks.rs#L294C1-L294C1].
+
+## Minimal Viable Prototype
+
+1. The user should be able to create the integration from Datadog in a couple of clicks.
```

Comment:
Do we need this for v1? The console integration seems like something we could layer on later, and not tackling it first means no fighting with OAuth.

Reply:
I'll revamp much of the content related to OAuth and the onboarding depending on our decision in Slack. In the meantime, I'm going to tackle the other comments.